lib/PuppeteerSharp/ChromeTargetManager.cs

```diff
@@ -35,8 +38,21 @@ internal class ChromeTargetManager : ITargetManager
             _targetFactoryFunc = targetFactoryFunc;
             _logger = _connection.LoggerFactory.CreateLogger<ChromeTargetManager>();
             _connection.MessageReceived += OnMessageReceived;
-            _connection.SessionDetached += Connection_SessionDetached;
+            _connection.SessionDetached += Connection_SessionDetached;
+        }
+
+        public event EventHandler<TargetChangedArgs> TargetAvailable;
+
+        public event EventHandler<TargetChangedArgs> TargetGone;
+
+        public event EventHandler<TargetChangedArgs> TargetChanged;
 
+        public event EventHandler<TargetChangedArgs> TargetDiscovered;
+
+        public ConcurrentDictionary<string, Target> GetAvailableTargets() => _attachedTargetsByTargetId;
+
+        public async Task InitializeAsync()
+        {
             _ = _connection.SendAsync("Target.setDiscoverTargets", new TargetSetDiscoverTargetsRequest
```

Comment:
Is there a benefit of moving the call to `Target.setDiscoverTargets` in `InitializeAsync` instead of having it in the constructor as it is in [puppeteer](https://github.com/puppeteer/puppeteer/blob/793a37183750b85ee71c8a6ac9d24f8e4ac0dff7/packages/puppeteer-core/src/common/ChromeTargetManager.ts#L103-L109)?

Reply:
I didn't fully understand this, so I'll have to run more tests to get a better grip at it. But the test PuppeteerConnectTests.ShouldSupportTargetFilter was failing randomly. This test connects to a browser that already has 3 tabs open.

What I found was that the messages of the targets were being processed concurrently, even before the event handlers in `AttachAsync` were added. Again, ultimately this is due to the multi-threaded .NET vs single-threaded JS difference, as in JS the constructor and the event handlers would be added without any concurrent code.

I will get back to you with a more thorough explanation.

Reply:
Another idea, if we move `Target.setDiscoverTargets` to `InitializeAsync` anyways, can we use modern async/await syntax and avoid `_targetDiscoveryCompletionSource` or I'm a missing some concurrency again?

```cs
public async Task InitializeAsync()
{
    var setDiscoverTargetsTask = _connection.SendAsync("Target.setDiscoverTargets", new TargetSetDiscoverTargetsRequest
    {
        Discover = true,
        Filter = new[]
        {
            new TargetSetDiscoverTargetsRequest.DiscoverFilter()
            {
                Type = "tab",
                Exclude = true,
            },
            new TargetSetDiscoverTargetsRequest.DiscoverFilter(),
        },
    });

    await _connection.SendAsync("Target.setAutoAttach", new TargetSetAutoAttachRequest()
    {
        WaitForDebuggerOnStart = true,
        Flatten = true,
        AutoAttach = true,
    }).ConfigureAwait(false);

    try
    {
        await setDiscoverTargetsTask.ConfigureAwait(false);
        StoreExistingTargetsForInit();
    }
    catch (Exception ex)
    {
        _logger.LogError(ex, "Target.setDiscoverTargets failed");
    }

    FinishInitializationIfReady();

    await _initializeCompletionSource.Task.ConfigureAwait(false);
}
```

Reply:
@jnyrup I have reverted these changes and couldn't reproduce any failure in 100 executions of PuppeteerConnectTests.ShouldSupportTargetFilter. Let's see how it behaves in AppVeyor.

Reply:
Turns out awaiting for `_targetDiscoveryCompletionSource` is needed before calling `FinishInitializationIfReady()`, otherwise `FinishInitializationIfReady` could execute before any targets have been added to the `_targetsIdsForInit` collection and the initialization would be considered completed.

Reply:
I'm fine with how it's done now 👍 

... but if we moved `_connection.SendAsync("Target.setDiscoverTargets"` to `InitializeAsync` again and awaited it, like in my example above, I'm failing to see how we can reach `FinishInitializationIfReady` before `setDiscoverTargetsTask` has completed, which should remove the need for `_targetDiscoveryCompletionSource`.

Reply:
@jnyrup had a closer look at your code above.

The change seems fine, and it simplifies a few things - for example it uses a try/catch which is more natural than checking `t.IsFaulted`. It's also closer to how FirefoxTargetManager works.

However, the `_targetDiscoveryCompletionSource` is still needed, to signal other threads that may be processing a message and executing `OnAttachedToTarget`.

It also has the downside of diverging from upstream.

I can commit these changes if you agree it's a better solution.

Reply:
Ahh, it's those events messing with my brain again...
I now (again) see why `_targetDiscoveryCompletionSource` still is necessary.

I'll let kblok decide whether to call `_connection.SendAsync("Target.setDiscoverTargets"` in the constructor of `InitializeAsync`.